File: 백준/Silver/17199. Milk Factory/Milk Factory.py
from collections import deque
N = int(input())
graph = [[] for _ in range(N+1)]
for _ in range(N-1):
    a, b = map(int, input().split())
    graph[b].append(a)
find = False
for i in range(1, N+1):
    visited = [False] * (N+1)
    visited[0] = True
    q = deque()
    q.append(i)
    visited[i] = True
    while q:
        now = q.popleft()
        for next_node in graph[now]:
            if not visited[next_node]:
                q.append(next_node)
                visited[next_node] = True
    if False not in visited:
        answer = i
        find = True
        break

if find:
    print(answer)
else:
    print(-1)

File: 백준/Silver/17199. Milk Factory/Milk Factory.java
import java.util.*;
import java.io.*;

public class Main {
    static int N;
    static ArrayList<ArrayList<Integer>> graph;
    static boolean[] visited;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        graph = new ArrayList<>();
        for (int i = 0; i <= N; i++) {
            graph.add(new ArrayList<>());
        }
        for (int i = 0; i < N-1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            graph.get(b).add(a);
        }
        boolean find = false;
        int answer = 0;
        for (int i = 1; i <= N; i++) {
            visited = new boolean[N+1];
            visited[0] = true;
            visited[i] = true;
            dfs(i);
            int cnt = 0;
            for (int j = 0; j <= N; j++) {
                if (visited[j]) {
                    cnt++;
                }

            }
            if (cnt == N+1) {
                find = true;
                answer = i;
                break;
            }
        }
        if (find) {
            System.out.println(answer);
        } else {
            System.out.println(-1);
        }


    }
    static void dfs(int now) {
        for (int next : graph.get(now)) {
            if (!visited[next]) {
                visited[next] = true;
                dfs(next);
            }
        }
    }
}
